### ISSUE TITLE
Support versioning dataset annotations and metadata (maybe also images)

### ISSUE THREAD

**Issue by jnothman** [Weed-AI/Weed-AI#591]
The repository structure should be changed to support tagged versions (and perhaps a mutable/unstable head version), such that:
* all tagged versions are retained, ideally without files duplicated on disk
* a zip of each version is available for download
* a dataset page can be rendered for each version (?) and the user can switch between versions in the dataset page
* the latest stable version is rendered by default at the existing endpoints
* the weedcoco.json (ideally) or another file in the repository includes an indicator of version within the metadata

**Comment by geezacoleman** [comment]
Nice to see this coming together! Would there be a change log for each version, similar to GitHub's release/versioning tools? I imagine free text entry would be enough, just so the person could describe what's new/different.

**Comment by jnothman** [comment]
We scoped the idea of "work in progress" datasets as entailing the need for versioning, so this was inevitable, but we've realised it's something we can already tackle soon.

We hadn't spoken about commit messages / release notes, but I suppose that's a very appropriate thing to support/require.

**Comment by spikelynch** [comment]
Here are my initial thoughts on the basic operations we need to support versioning. Assumptions:

* a dataset is associated with the user who created it, and only a dataset's owner can update it.
* we want to preserve current features like checking for duplicate images

Minimal operations for versioning:

* create - add a collection of images, with weedcoco.json (which may be crosswalked from COCO, VOC or CVAT) as a new dataset

* index - extract all the information required for the search from a dataset and add it to the index. This will include all the info we're currently indexing, plus information about the dataset's versions, and also possibly image hashes (see note on deduplication below).

* update - edit a dataset by adding a new version - this new version can be edits to the metadata, new annotations, new images (or removal of images)

* compare - for a new version, compare it against the previous version (to avoid adding the same files multiple times) and against the rest of the repository (to avoid duplicating images)

* export - create a zipfile corresponding to a particular version of a dataset

Optional features:

* commit - make the current draft version of a dataset "permanent" (it can still be updated by a subsequent update)

* export all - create a zipfile with the dataset's entire history

Not all of these operations will need to be user-visible in a single release. For example, we could implement versioning as a way to support the edit functionality, without needing to expose multiple versions of a dataset in the search interface (ie the only visible version of a dataset would be the latest) and add the ability to access versions in a later release.

Regarding deduplication, if I've understood the current code, when we add a dataset, it traverses the repository and generates hashes for everything and compares that to hashes for the incoming images.

OCFL (one of the technology options we're considering for versioning) creates hashes for all content - we could leverage this and add those hashes to the index so that checking for duplicates would be more scalable. Duplicate detection will need to be modified to accommodate updating a dataset (because we don't want to reject a new version for having the same images as the old version).

Questions

1. Is maintaining a 'draft' head of a dataset something we want to be done by the versioning system? Or is it something we could better manage using an upload status flag and keep the versioning simpler?

2. The system currently stores some metadata in both the index and the postgres database. I've assumed that the indexing process above would also put what's required into the database.

**Comment by jnothman** [comment]
Thanks for this!

Re compare: I don't think we're going to facilitate this in any interesting way. We could allow a user to download all versions of a dataset as OCFL, however, leaving the comparison up to them.

> it traverses the repository and generates hashes for everything and compares that to hashes for the incoming images.

We actually rename the image files to their hash... which might not be ideal from the perspective of preserving what the user uploaded, but it's what we do!

> we could leverage this and add those hashes to the index so that checking for duplicates would be more scalable. 

The current image hashing is using a hash that will create collisions for near-duplicate images (i.e. a locality-sensitive hash). I wonder what the benefit is of changing the current hashing scheme?

The primary queries that remain for me relate to the draft status and what aspects of non-current versions need to be accessed, which you don't really discuss much. That is, we might think of how users want to access different versions:

|                                                         | Latest release | Archival releases | Draft before 1st release | Draft edits upon latest release (do we want to support this?)      |
| ------------------------------------------------------- | -------------- | ----------------- | ------------------------ | ------------------------------------------------------------------ |
| Can retrieve Zip                                        | Yes            | Yes               | Maybe                    | Maybe                                                              |
| Can get link to Zip                                     | Yes            | Probably          | Maybe                    | Maybe                                                              |
| Can see in Explore                                      | Yes            | No                | Probably                 | Probably not                                                       |
| Listed in Datasets                                      | Yes            | No                | Yes                      | Maybe (indicates that there are updates and invites contributions) |
| Dataset summary page available                          | Yes            | Maybe             | Yes                      | Maybe                                                              |
| Images and bounding boxes shown in dataset summary page | Yes            | Probably not      | Probably                 | Probably not                                                       |

Hmmm this is also assuming members of the public are doing the above, though you might want to consider different access control for the creator/admin.

Re draft status / mutable head, how important do you think this is prior to the first release?

**Comment by spikelynch** [comment]
> We actually rename the image files to their hash... which might not be ideal from the perspective of preserving what the user uploaded, but it's what we do!

ah -ok, that makes more sense.

> The current image hashing is using a hash that will create collisions for near-duplicate images (i.e. a locality-sensitive hash). I wonder what the benefit is of changing the current hashing scheme?

The OCFL spec (and thus also the libraries) require either sha256 or sha512 (see https://ocfl.io/1.0/spec/#digests for details)

There's no benefit other than compliance to the spec, and we'd lose the near-duplicate feature of the current hash if we just use the OCFL digests.

We can still generate the current hashes and store them in the same way - we'll need to do something about how duplicate detection works in any case.

> The primary queries that remain for me relate to the draft status and what aspects of non-current versions need to be accessed, which you don't really discuss much. That is, we might think of how users want to access different versions:

Thanks for this - I'll have a think about it.
 
> Re draft status / mutable head, how important do you think this is prior to the first release?

I don't think that it's a high priority, although there might be ways of implementing it without complicating the versioning too much - for instance the upload dir could be where the draft lives until it's time to commit it.

**Comment by jnothman** [comment]
> for instance the upload dir could be where the draft lives until it's time to commit it.

Except that we want to display at least a placeholder for it, ideally with statistics if not image samples, in Datasets?

**Comment by spikelynch** [comment]
Here is a draft of a model for versioning updates. I've used the term 'release' for any dataset which is visible in the index and web interface, whether it's an under-construction snapshot or a published / 'final' version (scare quotes because even final versions need metadata corrections).

The idea is that the distinction between under-construction or published is something we should store in the metadata for the dataset, rather than be a state of the versioning system, so that the latter can be as simple as possible.

The system would build a sequence of edits as ocfl versions (v1, v2, v3 etc). Once a version is approved it gets updated in ocfl (so that the release status and message can be stored back in the repo) and indexed.

The three actions from the point of view of the backend are:

* write this set of edits to a dataset
* approve this version of the dataset (which is a special kind of write)
* index this version of the dataset as a release

![versioning drawio](https://user-images.githubusercontent.com/3653710/155033762-a6bf72b2-159b-4dda-8662-dafd88d7d1c7.svg)

**Comment by spikelynch** [comment]
[State diagram for the above](https://bit.ly/3vbXSvv)
![version_states](https://user-images.githubusercontent.com/3653710/155034708-26f8ee8e-511c-4f53-984e-7edb3c5bf8c8.svg)

**Comment by spikelynch** [comment]
Clarification of the above - the v1, v2, v3 versions on the left represent updates of the dataset as a whole - ie the operation which happens at the end of the stepper - not the individual writes which take place during the stepper (for instance, writing the agcontext).

So this design still requires an upload scratch space in which a dataset is built, and which then gets committed to the repo as a new version.

**Comment by jnothman** [comment]
I had imagined datasets under construction needing _some_ kind of review before being visible. But that's for later...

**Comment by spikelynch** [comment]
That's still possible in this model. What I've called a "release" (but we should maybe rename) is any visible dataset, so there's always a review stage before visibility.

How I imagine us handling under-construction v published is as a field in the metadata, so there are different sorts of released datasets, but the same versioning workflow underneath them.

One way of thinking of this is that it decouples versioning states from workflow states, and managing the workflow states (ie - a release is 'under construction' then 'published' - or directly to published - etc) is a separate concern. Which I'm deferring for now, and which we can change independently of how versioning works.

**Comment by geezacoleman** [comment]
This is all looking really great @spikelynch! One question I hopefully haven't missed in this thread - would edits made prior to the successful review be released in the first public version (v1)? Probably makes sense to start any versioning from v1, which I'm guessing is what happens here.

**Comment by jnothman** [comment]
> What I've called a "release" (but we should maybe rename) is any visible dataset, so there's always a review stage before visibility.

Yes, and we could still have logic to skip review for subsequent updates to a WIP dataset.

> would edits made prior to the successful review be released in the first public version (v1)? 

the version numbers here are for the OCFL machinery, not the users. There would be a mapping from release version names to OCFL version numbers. For a work in progress dataset you might imagine this in the WeedCOCO metadata:
```
"version": "pre-release",
"history": [],
```
For a released dataset you might have:
```
"version": "2",
"history": [
    {"version": "2", "ocfl_version": 12, "url": "https://weed-ai.sydney.edu.au/datasets/a1b2c3;version=2"},
    {"version": "1", "ocfl_version": 7, "url": "https://weed-ai.sydney.edu.au/datasets/a1b2c3;version=1"}
]
```
Unpublished versions would still be stored in the OCFL-based repository, but wouldn't ordinarily be accessible to users.

**Comment by jnothman** [comment]
Think of the v1, v2, v3, more like Git's commit ref SHAs.

**Comment by spikelynch** [comment]
Exactly - the public versions of datasets (whether under-construction or 'final') will have their own increasing versions (in the diagram I've called them 'release 1', 'release 2').

Having two versioning schemes gives us a bit of flexibility in how we pick a convention for public-facing version numbers, and means that these aren't dependent on the backend platform.
